src/main/scala/inox/solvers/combinators/NonIncrementalSolver.scala

```diff
@@ -0,0 +1,49 @@
+/* Copyright 2009-2018 EPFL, Lausanne */
+
+package inox
+package solvers
+package combinators
+
+import utils._
+import scala.concurrent.duration._
+import scala.collection.mutable.ArrayBuffer
+
+trait NonIncrementalSolver extends AbstractSolver { self =>
+  import program.trees._
+  import SolverResponses._
+
+  protected def underlying(): AbstractSolver {
+    val program: self.program.type
+    type Trees = self.Trees
+    type Model = self.Model
+  }
+
+  val assertions: IncrementalSet[Trees] = new IncrementalSet[Trees]
+
+  val allSolvers: ArrayBuffer[AbstractSolver] = ArrayBuffer()
+
+  def assertCnstr(expression: Trees): Unit = assertions += expression
+  def reset(): Unit = assertions.clear()
+  def free(): Unit = for (solver <- allSolvers) solver.free()
+  def interrupt(): Unit = for (solver <- allSolvers) solver.interrupt()
+
+  override def check(config: CheckConfiguration): config.Response[Model, Assumptions] = {
+    val newSolver = underlying()
+    allSolvers.append(newSolver)
+    for (expression <- assertions)
+      newSolver.assertCnstr(expression)
+    newSolver.check(config)
```

Comment:
Don't you think we should free the solver once it finishes `check`? Why are you keeping multiple solvers around?

Reply:
You're right, can we assume that check's are called sequentially? So I keep an `Option[Solver]` to `free` or `interrupt` it if necessary, and then I remove it at the end of the check.

Reply:
I'm not sure what you mean by sequentially. In general, I don't think we support multi-threading on the solvers. I would rather do something like
```scala
val newSolver = underlying()
try {
  ...
  newSolver.check(config)
} finally {
  newSolver.free()
}
```

Reply:
I also keep the solver in the `currentSolver` option to interrupt it, right?

Reply:
Oooh, I hadn't realized you were keeping it around for interrupt :) my bad!

Yes, you should indeed keep it around in that case. Maybe just to keep things clean, you could throw an exception from `check` and `checkAssumptions` if the `currentSolver` isn't empty. (Or keep solvers in a list and add synchronization, but I don't see any reason to support multi-threading on this solver when it's not available on the others anyway.)

Reply:
I added an assertion, thanks